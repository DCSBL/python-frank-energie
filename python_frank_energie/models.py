"""Data models for the Frank Energie API."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from dateutil import parser

from .exceptions import AuthException, RequestException

_LOGGER = logging.getLogger(__name__)


@dataclass
class Authentication:
    """Authentication data.

    Generated by the login or renewToken mutation.

    authToken: The token to use for authenticated requests.
    refreshToken: The token to use to renew the authToken.
    """

    authToken: str
    refreshToken: str

    @staticmethod
    def from_dict(data: dict[str, str]) -> Authentication:
        """Parse the response from the login or renewToken mutation."""
        _LOGGER.debug("Authentication %s", data)

        if errors := data.get("errors"):
            raise AuthException(errors[0]["message"])

        login_payload = data.get("data", {}).get("login")
        renew_payload = data.get("data", {}).get("renewToken")
        if not login_payload and not renew_payload:
            raise AuthException("Unexpected response")

        payload = login_payload or renew_payload

        return Authentication(
            authToken=payload.get("authToken"),
            refreshToken=payload.get("refreshToken"),
        )


@dataclass
class Invoices:
    """Invoices data, including the previous, current and upcoming period."""

    @dataclass
    class Invoice:
        """Invoice data, including the start date, period description and total amount."""

        StartDate: datetime
        PeriodDescription: str
        TotalAmount: float

        @staticmethod
        def from_dict(data: dict[str, str]) -> Invoices.Invoice | None:
            """Parse the response from the invoices query."""
            if data is None:
                return None

            return Invoices.Invoice(
                StartDate=parser.parse(data.get("StartDate")),
                PeriodDescription=data.get("PeriodDescription"),
                TotalAmount=data.get("TotalAmount"),
            )

    previousPeriodInvoice: Invoice | None
    currentPeriodInvoice: Invoice | None
    upcomingPeriodInvoice: Invoice | None

    @staticmethod
    def from_dict(data: dict[str, str]) -> Invoices:
        """Parse the response from the invoices query."""
        _LOGGER.debug("Invoices %s", data)

        if errors := data.get("errors"):
            raise RequestException(errors[0]["message"])

        payload = data.get("data", {}).get("invoices")
        if not payload:
            raise RequestException("Unexpected response")

        return Invoices(
            previousPeriodInvoice=Invoices.Invoice.from_dict(
                payload.get("previousPeriodInvoice")
            ),
            currentPeriodInvoice=Invoices.Invoice.from_dict(
                payload.get("currentPeriodInvoice")
            ),
            upcomingPeriodInvoice=Invoices.Invoice.from_dict(
                payload.get("upcomingPeriodInvoice")
            ),
        )


@dataclass
class User:
    """User data, including the current status of the connection."""

    connectionsStatus: str
    firstMeterReadingDate: str
    lastMeterReadingDate: str
    advancedPaymentAmount: float
    hasCO2Compensation: bool

    @staticmethod
    def from_dict(data: dict[str, str]) -> User:
        """Parse the response from the me query."""
        _LOGGER.debug("User %s", data)

        if errors := data.get("errors"):
            raise RequestException(errors[0]["message"])

        payload = data.get("data", {}).get("me")
        if not payload:
            raise RequestException("Unexpected response")

        return User(
            connectionsStatus=payload.get("connectionsStatus"),
            firstMeterReadingDate=payload.get("firstMeterReadingDate"),
            lastMeterReadingDate=payload.get("lastMeterReadingDate"),
            advancedPaymentAmount=payload.get("advancedPaymentAmount"),
            hasCO2Compensation=payload.get("hasCO2Compensation"),
        )


@dataclass
class MonthSummary:
    """Month summary data, including the actual and expected costs for this month."""

    actualCostsUntilLastMeterReadingDate: float
    expectedCostsUntilLastMeterReadingDate: float
    expectedCosts: float
    lastMeterReadingDate: str

    @staticmethod
    def from_dict(data: dict[str, str]) -> MonthSummary:
        """Parse the response from the monthSummary query."""
        _LOGGER.debug("MonthSummary %s", data)

        if errors := data.get("errors"):
            raise RequestException(errors[0]["message"])

        payload = data.get("data", {}).get("monthSummary")
        if not payload:
            raise RequestException("Unexpected response")

        return MonthSummary(
            actualCostsUntilLastMeterReadingDate=payload.get(
                "actualCostsUntilLastMeterReadingDate"
            ),
            expectedCostsUntilLastMeterReadingDate=payload.get(
                "expectedCostsUntilLastMeterReadingDate"
            ),
            expectedCosts=payload.get("expectedCosts"),
            lastMeterReadingDate=payload.get("lastMeterReadingDate"),
        )

    @property
    def differenceUntilLastMeterReadingDate(self):
        """The difference between the expected costs and the actual costs."""
        return round(
            self.actualCostsUntilLastMeterReadingDate
            - self.expectedCostsUntilLastMeterReadingDate,
            2,
        )


class Price:
    """Price data for one hour."""

    date_from: datetime
    date_till: datetime
    market_price: float
    market_price_tax: float
    sourcing_markup_price: float
    energy_tax_price: float

    def __init__(self, data: dict) -> None:
        """Parse the response from the prices query."""
        self.date_from = parser.parse(data["from"])
        self.date_till = parser.parse(data["till"])

        self.market_price = data["marketPrice"]
        self.market_price_tax = data["marketPriceTax"]
        self.sourcing_markup_price = data["sourcingMarkupPrice"]
        self.energy_tax_price = data["energyTaxPrice"]

    def __str__(self) -> str:
        """Return a string representation of this price entry."""
        return f"{self.date_from} -> {self.date_till}: {self.total}"

    @property
    def for_now(self) -> bool:
        """Whether this price entry is for the current hour."""
        return self.date_from <= datetime.now(timezone.utc) < self.date_till

    @property
    def for_future(self) -> bool:
        """Whether this price entry is for and hour after the current one."""
        return self.date_from.hour > datetime.now(timezone.utc).hour

    @property
    def for_today(self) -> bool:
        """Whether this price entry is for the current day."""
        day_start = datetime.now(timezone.utc).replace(
            hour=0, minute=0, second=0, microsecond=0
        )
        day_end = day_start + timedelta(days=1)
        return self.date_from >= day_start and self.date_till <= day_end

    @property
    def market_price_with_tax(self) -> float:
        """The market price including tax."""
        return round(self.market_price + self.market_price_tax, 4)

    @property
    def total(self) -> float:
        """The total price for this hour."""
        return round(
            self.market_price
            + self.market_price_tax
            + self.sourcing_markup_price
            + self.energy_tax_price,
            4,
        )


class PriceData:
    """Price data for a period of time."""

    price_data: list[Price] = []

    def __init__(self, price_data: list[dict] | None = None) -> None:
        """Parse the response from the prices query."""
        if price_data is not None:
            self.price_data = [Price(price) for price in price_data]

    def __add__(self, b: PriceData) -> PriceData:
        """Combine two PriceData objects."""
        pd = PriceData()
        pd.price_data = self.price_data + b.price_data
        return pd

    def __str__(self):
        """Return a string representation of this price data."""
        return str([str(price) for price in self.price_data])

    @property
    def all(self) -> list[Price]:
        """All prices."""
        return self.price_data

    @property
    def today(self) -> list[Price]:
        """Prices for today."""
        return [hour for hour in self.price_data if hour.for_today]

    @property
    def current_hour(self) -> Price:
        """Price that's currently applicable."""
        return [hour for hour in self.price_data if hour.for_now][0]

    @property
    def today_min(self) -> Price:
        """Price with the lowest total for today."""
        return min([hour for hour in self.today], key=lambda hour: hour.total)

    @property
    def today_max(self) -> Price:
        """Price with the highest total for today."""
        return max([hour for hour in self.today], key=lambda hour: hour.total)

    @property
    def today_avg(self) -> float:
        """Average price for today."""
        return round(sum(hour.total for hour in self.today) / len(self.today), 5)

    def get_future_prices(self) -> list[Price]:
        """Prices for hours after the current one."""
        return [hour for hour in self.price_data if hour.for_future]

    def asdict(self, attr) -> dict:
        """Return a dict that can be used as entity attribute data."""
        return [
            {"from": e.date_from, "till": e.date_till, "price": getattr(e, attr)}
            for e in self.price_data
        ]


@dataclass
class MarketPrices:
    """Market prices for electricity and gas."""

    electricity: PriceData
    gas: PriceData

    @staticmethod
    def from_dict(data: dict[str, str]) -> MarketPrices:
        """Parse the response from the marketPrices query."""
        _LOGGER.debug("Prices %s", data)

        if errors := data.get("errors"):
            if errors[0]["message"].startswith("No marketprices found for segment"):
                return MarketPrices(PriceData(), PriceData())

            raise RequestException(errors[0]["message"])

        payload = data.get("data")
        if not payload:
            raise RequestException("Unexpected response")

        return MarketPrices(
            electricity=PriceData(payload.get("marketPricesElectricity")),
            gas=PriceData(payload.get("marketPricesGas")),
        )

    @staticmethod
    def from_userprices_dict(data: dict[str, str]) -> MarketPrices:
        """Parse the response from the marketPrices query."""
        _LOGGER.debug("Prices %s", data)

        if errors := data.get("errors"):
            if errors[0]["message"].startswith("No marketprices found for segment"):
                return MarketPrices(PriceData(), PriceData())

            raise RequestException(errors[0]["message"])

        payload = data.get("data")
        if not payload:
            raise RequestException("Unexpected response")

        customerMarketPrices = payload.get("customerMarketPrices")

        return MarketPrices(
            electricity=PriceData(customerMarketPrices.get("electricityPrices")),
            gas=PriceData(customerMarketPrices.get("gasPrices")),
        )
