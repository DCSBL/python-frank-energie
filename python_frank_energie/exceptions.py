"""Exceptions for the Frank Energie API."""


class FrankEnergieException(Exception):
    """Base exception."""


class AuthRequiredException(FrankEnergieException):
    """Authentication required for this request."""


class AuthException(FrankEnergieException):
    """Authentication/login failed."""


class RequestException(FrankEnergieException):
    """Request failed."""
