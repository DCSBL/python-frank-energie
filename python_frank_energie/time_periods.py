from enum import Enum
# time_periods.py

class TimePeriod(Enum):
    """
    Enum representing different time periods that can be used to filter data.

    Usage:
    - `TimePeriod.PREVIOUS_HOUR` represents the hour before the current hour.
    - `TimePeriod.CURRENT_HOUR` represents the current hour.
    - `TimePeriod.NEXT_HOUR` represents the hour after the current hour.
    - `TimePeriod.YESTERDAY` represents the previous day.
    - `TimePeriod.TODAY` represents the current day.
    - `TimePeriod.TOMORROW` represents the next day.
    - `TimePeriod.UPCOMING` represents all future hours.
    - `TimePeriod.BEFORE_6AM` represents all hours before 6 am of the current day.
    - `TimePeriod.AFTER_6AM` represents all hours after 6 am of the current day.
    - `TimePeriod.TOMORROW_BEFORE_6AM` represents all hours before 6 am of the next day.
    - `TimePeriod.TOMORROW_AFTER_6AM` represents all hours after 6 am of the next day.
    - `TimePeriod.PREVIOUS_MONTH` represents all invoices for the previous month.
    - `TimePeriod.CURRENT_MONTH` represents all invoices for the current month.
    - `TimePeriod.NEXT_MONTH` represents all invoices for the next month.
    - `TimePeriod.CURRENT_YEAR` represents all invoices for the current year.
    - `TimePeriod.PREVIOUS_YEAR` represents all invoices for the previous year.
    - `TimePeriod.ALLTIME` represents all prices or invoices.
    """

    PREVIOUS_HOUR = 'previous_hour'  # The hour before the current hour.
    CURRENT_HOUR = 'current_hour'  # The current hour.
    NEXT_HOUR = 'next_hour'  # The hour after the current hour.
    YESTERDAY = 'yesterday'  # The previous day.
    TODAY = 'today'  # The current day.
    TOMORROW = 'tomorrow'  # The next day.
    UPCOMING = 'upcoming'  # All future hours.
    BEFORE_6AM = 'before_6am'  # All hours before 6am of the current day.
    AFTER_6AM = 'after_6am'  # All hours after 6am of the current day.
    TOMORROW_BEFORE_6AM = 'tomorrow_before_6am'  # All hours before 6am of the next day.
    TOMORROW_AFTER_6AM = 'tomorrow_after_6am'  # All hours after 6am of the next day.
    PREVIOUS_MONTH = 'previous_month'  # All invoices for previous month.
    CURRENT_MONTH = 'current_month'  # All invoices for current month.
    NEXT_MONTH = 'next_month'  # All invoices for previous month.
    CURRENT_YEAR = 'current_year'  # All invoices for current year.
    PREVIOUS_YEAR = 'previous_year'  # All invoices for previous year.
    ALLTIME = "all_time" # All prices or invoices

    def __str__(self):
        return self.value

    #not in use    
    def get_prices_for_time_period(self, price_data, time_period: 'TimePeriod'):
        if time_period == TimePeriod.TODAY:
            return [hour for hour in price_data if hour.for_today]
        elif time_period == TimePeriod.TOMORROW:
            return [hour for hour in price_data if hour.for_tomorrow]
        else:
            raise ValueError(f"Invalid time period: {time_period}")
