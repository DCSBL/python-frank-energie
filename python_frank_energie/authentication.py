# python_frank_energie/authentication.py

import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional

import jwt
from .exceptions import AuthException

_LOGGER: logging.Logger = logging.getLogger(__name__)

@dataclass
class Authentication:
    """Authentication data.

    Generated by the login or renewToken mutation.

    authToken: The token to use for authenticated requests.
    refreshToken: The token to use to renew the authToken.
    """

    authToken: str
    refreshToken: str

    @staticmethod
    def from_dict(data: dict[str, str]) -> 'Authentication':
        """Parse the response from the login or renewToken mutation."""
        _LOGGER.debug("Authentication %s", data)

        if errors := data.get("errors"):
            raise AuthException(errors[0]["message"])

        login_payload = data.get("data", {}).get("login")
        renew_payload = data.get("data", {}).get("renewToken")
        if not login_payload and not renew_payload:
            raise AuthException("Unexpected response")

        payload = Authentication._extract_payload(data)
        # if not payload:
        #     raise AuthException("Unexpected response")
        if not payload or "authToken" not in payload or "refreshToken" not in payload:
            raise AuthException("Unexpected response: Missing or incomplete payload")

        return Authentication(
            # authToken=payload.get("authToken"),
            # refreshToken=payload.get("refreshToken"),
            authToken=payload["authToken"],
            refreshToken=payload["refreshToken"]
        )

    def __init__(self, authToken: str, refreshToken: str):
        self.authToken = authToken
        self.refreshToken = refreshToken

    @staticmethod
    def _extract_payload(data: dict) -> Optional[dict]:
        """Extract the login or renewToken payload from the data dictionary."""
        return data.get("data", {}).get("login") or data.get("data", {}).get("renewToken")

    def authTokenValid(self, tz: timezone = timezone.utc) -> bool:
        """Return that authToken is valid according to expiration time."""
        authTokenDecoded = jwt.decode(
            self.authToken,
            verify=True,
            algorithms=["HS256"],
            options={"verify_signature": False},
        )
        return datetime.fromtimestamp(
            authTokenDecoded["exp"], tz=timezone.utc
        ) > datetime.now(tz=tz)

    def __repr__(self):
        return f"Authentication(authToken={self.authToken}, refreshToken={self.refreshToken})"

class AuthenticationResult:
    def __init__(self, authToken: str, refreshToken: str):
        self.authToken = authToken
        self.refreshToken = refreshToken