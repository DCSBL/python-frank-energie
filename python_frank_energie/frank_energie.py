"""FrankEnergie API implementation."""

from __future__ import annotations

import asyncio
from datetime import date
from typing import Any

from aiohttp.client import ClientError, ClientSession

from .exceptions import AuthException, AuthRequiredException
from .models import Authentication, Invoices, MarketPrices, MonthSummary, User


class FrankEnergie:
    """FrankEnergie API."""

    DATA_URL = "https://frank-graphql-prod.graphcdn.app/"

    def __init__(
        self,
        clientsession: ClientSession = None,
        auth_token: str | None = None,
        refresh_token: str | None = None,
    ):
        """Initialize the FrankEnergie client."""
        self._close_session: bool = False
        self._auth: Authentication | None = None
        self._session = clientsession

        if auth_token is not None or refresh_token is not None:
            self._auth = Authentication(auth_token, refresh_token)

    async def _query(self, query):
        if self._session is None:
            self._session = ClientSession()
            self._close_session = True

        try:
            resp = await self._session.post(
                self.DATA_URL,
                json=query,
                headers={"Authorization": f"Bearer {self._auth.authToken}"}
                if self._auth is not None
                else None,
            )

            response = await resp.json()

        except (asyncio.TimeoutError, ClientError, KeyError) as error:
            raise ValueError(f"Request failed: {error}") from error

        # Catch common error messages and raise a more specific exception
        if errors := response.get("errors"):
            for error in errors:
                if error["message"] == "user-error:auth-not-authorised":
                    raise AuthException

        return response

    async def login(self, username: str, password: str) -> Authentication:
        """Login and get the authentication token."""
        query = {
            "query": """
                mutation Login($email: String!, $password: String!) {
                    login(email: $email, password: $password) {
                        authToken
                        refreshToken
                    }
                }
            """,
            "operationName": "Login",
            "variables": {"email": username, "password": password},
        }

        self._auth = Authentication.from_dict(await self._query(query))
        return self._auth

    async def renew_token(self) -> Authentication:
        """Renew the authentication token."""
        if self._auth is None:
            raise AuthRequiredException

        query = {
            "query": """
                mutation RenewToken($authToken: String!, $refreshToken: String!) {
                    renewToken(authToken: $authToken, refreshToken: $refreshToken) {
                        authToken
                        refreshToken
                    }
                }
            """,
            "operationName": "RenewToken",
            "variables": {
                "authToken": self._auth.authToken,
                "refreshToken": self._auth.refreshToken,
            },
        }

        self._auth = Authentication.from_dict(await self._query(query))
        return self._auth

    async def month_summary(self) -> MonthSummary:
        """Get month summary data."""
        if self._auth is None:
            raise AuthRequiredException

        query = {
            "query": """
                query MonthSummary {
                    monthSummary {
                        actualCostsUntilLastMeterReadingDate
                        expectedCostsUntilLastMeterReadingDate
                        expectedCosts
                        lastMeterReadingDate
                    }
                }
            """,
            "operationName": "MonthSummary",
            "variables": {},
        }

        return MonthSummary.from_dict(await self._query(query))

    async def invoices(self) -> Invoices:
        """Get invoices data.

        Returns a Invoices object, containing the previous, current and upcoming invoice.
        """
        if self._auth is None:
            raise AuthRequiredException

        query = {
            "query": """
                query Invoices {
                    invoices {
                        previousPeriodInvoice {
                            StartDate
                            PeriodDescription
                            TotalAmount
                        }
                        currentPeriodInvoice {
                            StartDate
                            PeriodDescription
                            TotalAmount
                        }
                        upcomingPeriodInvoice {
                            StartDate
                            PeriodDescription
                            TotalAmount
                        }
                    }
                }
            """,
            "operationName": "Invoices",
            "variables": {},
        }

        return Invoices.from_dict(await self._query(query))

    async def user(self) -> User:
        """Get user data."""
        if self._auth is None:
            raise AuthRequiredException

        query = {
            "query": """
                query Me {
                    me {
                        ...UserFields
                    }
                }
                fragment UserFields on User {
                    connectionsStatus
                    firstMeterReadingDate
                    lastMeterReadingDate
                    advancedPaymentAmount
                    treesCount
                    hasCO2Compensation
                }
            """,
            "operationName": "Me",
            "variables": {},
        }

        return User.from_dict(await self._query(query))

    async def prices(
        self, start_date: date, end_date: date | None = None
    ) -> MarketPrices:
        """Get market prices."""
        query_data = {
            "query": """
                query MarketPrices($startDate: Date!, $endDate: Date!) {
                    marketPricesElectricity(startDate: $startDate, endDate: $endDate) {
                       from
                       till
                       marketPrice
                       marketPriceTax
                       sourcingMarkupPrice
                       energyTaxPrice
                    }
                    marketPricesGas(startDate: $startDate, endDate: $endDate) {
                       from
                       till
                       marketPrice
                       marketPriceTax
                       sourcingMarkupPrice
                       energyTaxPrice
                    }
                }
            """,
            "variables": {"startDate": str(start_date), "endDate": str(end_date)},
            "operationName": "MarketPrices",
        }

        return MarketPrices.from_dict(await self._query(query_data))

    async def user_prices(self, start_date: date) -> MarketPrices:
        """Get customer market prices."""
        if self._auth is None:
            raise AuthRequiredException

        query_data = {
            "query": """
                query CustomerMarketPrices($date: String!) {
                    customerMarketPrices(date: $date) {
                        electricityPrices {
                            from
                            till
                            marketPrice
                            marketPriceTax
                            sourcingMarkupPrice: consumptionSourcingMarkupPrice
                            energyTaxPrice: energyTax
                        }
                        gasPrices {
                            from
                            till
                            marketPrice
                            marketPriceTax
                            sourcingMarkupPrice: consumptionSourcingMarkupPrice
                            energyTaxPrice: energyTax
                        }
                    }
                }
            """,
            "variables": {"date": str(start_date)},
            "operationName": "CustomerMarketPrices",
        }

        return MarketPrices.from_userprices_dict(await self._query(query_data))

    @property
    def is_authenticated(self) -> bool:
        """Return if client is authenticated.

        Does not actually check if the token is valid.
        """
        return self._auth is not None

    async def close(self) -> None:
        """Close client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> FrankEnergie:
        """Async enter.

        Returns:
            The FrankEnergie object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
