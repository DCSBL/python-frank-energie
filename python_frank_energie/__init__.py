"""Frank Energie API library."""
from .frank_energie import FrankEnergie
from .models import Price, PriceData

__all__ = [
    "FrankEnergie",
    "Price",
    "PriceData",
]
