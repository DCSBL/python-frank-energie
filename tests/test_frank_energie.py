"""Test for Frank Energie."""

from datetime import datetime

import aiohttp
import pytest
from syrupy.assertion import SnapshotAssertion

from python_frank_energie import FrankEnergie
from python_frank_energie.exceptions import AuthException, AuthRequiredException

from . import load_fixtures

SIMPLE_DATA_URL = "frank-graphql-prod.graphcdn.app"


@pytest.mark.asyncio
async def test_init_without_authentication():
    """Test init without authentication."""
    api = FrankEnergie()
    assert api.is_authenticated is False


@pytest.mark.asyncio
async def test_init_with_authentication():
    """Test init with authentication."""
    api = FrankEnergie(auth_token="a", refresh_token="b")  # noqa: S106
    assert api.is_authenticated is True


#
# Login tests
#


@pytest.mark.asyncio
async def test_login(aresponses):
    """Test login."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("authentication.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        auth = await api.login("a", "b")  # noqa: S106
        await api.close()

    assert api.is_authenticated is True
    assert auth.authToken == "hello"
    assert auth.refreshToken == "world"


@pytest.mark.asyncio
async def test_login_invalid_credentials(aresponses):
    """Test login with invalid credentials."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("response_with_error.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        with pytest.raises(AuthException):
            await api.login("a", "b")  # noqa: S106
        await api.close()


@pytest.mark.asyncio
async def test_login_invalid_response(aresponses):
    """Test login with invalid response."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text="{}",
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        with pytest.raises(AuthException):
            await api.login("a", "b")  # noqa: S106
        await api.close()


#
# RenewToken tests


@pytest.mark.asyncio
async def test_renew_token(aresponses):
    """Test login."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("authentication.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, "a", "b")  # noqa: S106
        auth = await api.renew_token()
        await api.close()

    assert api.is_authenticated is True
    assert auth.authToken == "hello"
    assert auth.refreshToken == "world"


@pytest.mark.asyncio
async def test_renew_token_invalid_credentials(aresponses):
    """Test login with invalid credentials."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("response_with_error.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, "a", "b")  # noqa: S106
        with pytest.raises(AuthException):
            await api.renew_token()
        await api.close()


@pytest.mark.asyncio
async def test_renew_token_invalid_response(aresponses):
    """Test login with invalid response."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text="{}",
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, "a", "b")  # noqa: S106
        with pytest.raises(AuthException):
            await api.renew_token()
        await api.close()


#
# Month Summary
#


@pytest.mark.asyncio
async def test_month_summary(aresponses, snapshot: SnapshotAssertion):
    """Test request with authentication."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("month_summary.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, auth_token="a", refresh_token="b")  # noqa: S106
        summary = await api.month_summary("1234AB 10")
        await api.close()

    assert summary is not None
    assert summary == snapshot


@pytest.mark.asyncio
async def test_month_summary_without_authentication(aresponses):
    """Test request without authentication.

    'month_summary' request requires authentication.
    """
    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        with pytest.raises(AuthRequiredException):
            await api.month_summary("1234AB 10")
        await api.close()


#
# Invoices
#


@pytest.mark.asyncio
async def test_invoices(aresponses, snapshot: SnapshotAssertion):
    """Test request with authentication."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("invoices.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, auth_token="a", refresh_token="b")  # noqa: S106
        invoices = await api.invoices("1234AB 10")
        await api.close()

    assert invoices is not None
    assert invoices == snapshot


@pytest.mark.asyncio
async def test_invoices_without_authentication(aresponses):
    """Test request without authentication.

    'invoices' request requires authentication.
    """
    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        with pytest.raises(AuthRequiredException):
            await api.invoices("1234AB 10")
        await api.close()


#
# Me
#


@pytest.mark.asyncio
async def test_me(aresponses, snapshot: SnapshotAssertion):
    """Test request with authentication."""
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("me.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, auth_token="a", refresh_token="b")  # noqa: S106
        me = await api.me("1234AB 10")
        await api.close()

    assert me is not None
    assert me == snapshot


@pytest.mark.asyncio
async def test_me_without_authentication(aresponses):
    """Test request without authentication.

    'user' request requires authentication.
    """
    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        with pytest.raises(AuthRequiredException):
            await api.me("1234AB 10")
        await api.close()


#
# Prices
#


@pytest.mark.asyncio
async def test_prices(aresponses):
    """Test request without authentication.

    'prices' request does not require authentication.
    """
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("market_prices.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session)
        prices = await api.prices(datetime.utcnow().date(), datetime.utcnow().date())
        await api.close()

    assert prices.electricity is not None
    assert len(prices.electricity.price_data) == 24

    assert prices.gas is not None
    assert len(prices.gas.price_data) == 24


@pytest.mark.asyncio
async def test_user_prices(aresponses):
    """Test request with authentication.

    'prices' request does not require authentication.
    """
    aresponses.add(
        SIMPLE_DATA_URL,
        "/",
        "POST",
        aresponses.Response(
            text=load_fixtures("customer_market_prices.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        api = FrankEnergie(session, auth_token="a", refresh_token="b")  # noqa: S106
        prices = await api.user_prices(datetime.utcnow().date())
        await api.close()

    assert prices.electricity is not None
    assert len(prices.electricity.price_data) == 24

    assert prices.gas is not None
    assert len(prices.gas.price_data) == 24
