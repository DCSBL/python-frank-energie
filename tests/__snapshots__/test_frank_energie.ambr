# serializer version: 1
# name: test_invoices
  Invoices(previousPeriodInvoice=Invoices.Invoice(StartDate=datetime.datetime(2023, 3, 1, 0, 0), PeriodDescription='Maart 2023', TotalAmount=140.12), currentPeriodInvoice=Invoices.Invoice(StartDate=datetime.datetime(2023, 4, 1, 0, 0), PeriodDescription='April 2023', TotalAmount=80.34), upcomingPeriodInvoice=Invoices.Invoice(StartDate=datetime.datetime(2023, 5, 1, 0, 0), PeriodDescription='Mei 2023', TotalAmount=80.34))
# ---
# name: test_me
  Me(id='vb10h3v01nfb0vh1f', email='contact@email.nl', countryCode='NL', advancedPaymentAmount=100.0, treesCount=None, hasInviteLink=True, hasCO2Compensation=False)
# ---
# name: test_month_summary
  MonthSummary(actualCostsUntilLastMeterReadingDate=12.34, expectedCostsUntilLastMeterReadingDate=20.0, expectedCosts=50.0, lastMeterReadingDate='2023-01-01')
# ---
